### PR TITLE
API changes

### PR DESCRIPTION
**Pull request by rianadon**
I'm moving methods out of the `TED` class and into the spyder/mtu classes.

`reader.mtu_value(reader.mtus[2]).energy` becomes `reader.mtus[2].energy()`.

### CODE REVIEW

**Comment by rianadon** [comment]
@realumhelp I wanted to check with you on this one since it's changing quite a bit of the API & code.

**Review by realumhelp — APPROVED** [review]
I don't know what this does but I'm going to approve it! :smiley:

**Comment by realumhelp** [comment]
I looked through all the changes and I like them. I think this simplifies the interface and makes it easier to use. My only comment would be you may have lost a little bit of information on the type of data `energy()` is returning. So you should check the MTU's type at some point before using the results from `energy()` to know if it is a NET/LOAD/GEN/STAND-ALONE. I honestly don't really know the true difference between NET and LOAD, I think it just has to do with whether or not GEN is added to the total or not. For a system total display you wouldn't want to grab a STAND-ALONE MTU into that total because it has already been added in the NET MTU for instance.

It is late so hopefully that makes sense, the data is there just need to make sure it is interrupted properly.

**Comment by realumhelp** [comment]
This maybe future talk but when considering advanced system types that involve solar and such. Would you think `load()` and `generation()` type functions would be added similar to what I kinda brute forced here:

https://github.com/rianadon/the-energy-detective-py/pull/10/commits/c7a1d78301d27132aa0169dfd65004b0033dc9aa#diff-08c3daa550455abff6339903458fe845721453258e8bd9773a6396c2d22e35ffR117-R133

That data is specific to TED6000 so you have a small deviation between the TED5000 system.

**Comment by rianadon** [comment]
The `load()` and `generation()` sound great. I'm going to hold off on merging this until you finish that PR, then I can rebase and add those methods.

I don't understand the NET/GEN types either. I do agree though on the energy type; the returned energy type needs to be more specific.

**Comment by realumhelp** [comment]
> The `load()` and `generation()` sound great. I'm going to hold off on merging this until you finish that PR, then I can rebase and add those methods.
> 
> I don't understand the NET/GEN types either. I do agree though on the energy type; the returned energy type needs to be more specific.

My theory is only certain queries make sense for certain SystemTypes:
SystemType.NET: 
  * NET Query
SystemType.LOAD:
   * LOAD Query
   * GEN Query
 SystemLoad.GEN:
   * NET Query
   * GEN Query

**Comment by rianadon** [comment]
I'm a bit confused. Why does SystemType LOAD type have GEN Query? Or is that supposed to be NET?

**Comment by realumhelp** [comment]
> I'm a bit confused. Why does SystemType LOAD type have GEN Query? Or is that supposed to be NET?

Just a guess but when you step through the TED settings UI you can clearly see the 3 different system types. A non-solar normal system will be called NET, the only MTU types allowed are (NET and STAND-ALONE). A solar config that is bi-directional, meaning the solar feeds both the grid and your house for consumption, that is called NET-GEN and the only allowed MTU types are NET, GEN, and STAND-ALONE (no LOAD option). The last type (NET-LOAD) is your solar only directly feeds the grid, not an internal breaker, so your panel is only consumption. This type only allows LOAD, GEN, and STAND-ALONE (no NET option).

Allowed MTU types based on SystemTypes allowed in TED Config UI:
SystemType.NET:
---NET
---STAND-ALONE
SystemType.LOAD:
---LOAD 
---GEN
---STAND-ALONE
SystemLoad.GEN:
---NET Query
---GEN Query
---STAND-ALONE

**Comment by realumhelp** [comment]
Looks like the formatting on my table got jacked up let me try to retry it:

These are what I think are the valid queries for the various system types.
SystemType.NET:
---NET Query
SystemType.LOAD:
---LOAD Query
---GEN Query
SystemLoad.GEN:
---NET Query
---GEN Query

**Comment by realumhelp** [comment]
So theory would conclude that these are the definitions of the queries based on SystemType:

SystemType.NET:
---NET Query - Sum of all MTUs (only positive no GEN allowed in this config)
---LOAD Query - Empty (Can verify with my current system)
---GEN Query - Empty (Can verify with my current system)
SystemType.LOAD:
---NET Query - My guess is this would be empty?
---LOAD Query - Sum of all LOAD MTUs (does not subtract any GEN because that is directly feeding the grid)
---GEN Query - Sum of all GEN MTUs
SystemLoad.GEN:
---NET Query - Sum of all MTUs including all GEN so possibly negative if production is higher than consumption
---LOAD Query - My guess is this would be empty?
---GEN Query - Sum of all GEN MTUs negative number.

Again this is just my guess as I can't test it for real. We would need to find someone who had such a setup and would be willing to run some queries.

**Comment by rianadon** [comment]
Mine is SystemType.NET. I think it might be best to hold off on implementing these and adding a note in the README in case someone with a different system type is looking to use the library.

**Comment by rianadon** [comment]
I simplified YieldType to now have only 3 possible values: `NET`, `LOAD`, and `GEN` (i.e. removing the MTU and SPYDER_GROUP types). Does it make sense that the spyder/mtu values would be of type `NET`?

**Comment by realumhelp** [comment]
> Mine is SystemType.NET. I think it might be best to hold off on implementing these and adding a note in the README in case someone with a different system type is looking to use the library.

My SystemType is NET as well. I think it would be safe to implement at least NET and NET-GEN system types and kind of ignore the LOAD-GEN system type. This type of system to me would be commercial rather than residential, no one gets solar and not use the power generated. I would probably rename `LOAD` to Stand-Alone and I think everything else should fall into place. I would pull the queries for NET and for GEN, leaving the LOAD query alone for now until we figure out what it does.

> I simplified YieldType to now have only 3 possible values: `NET`, `LOAD`, and `GEN` (i.e. removing the MTU and SPYDER_GROUP types). Does it make sense that the spyder/mtu values would be of type `NET`?

It seems like there are three levels of the TED system: Overall System Totals (combined MTU values), Individual MTUs, and Spyders. So totaling all Spyders will always be <= their parent MTU values, totaling the MTUs will always be <= the DashData total values (ignoring generation for simplicity).

I would say all Spyder values are Stand-Alone, they aren't used for any of the totaling values. An MTU will be part of the NET total calculation so calling it NET could make sense.

**Comment by rianadon** [comment]
> I would probably rename LOAD to Stand-Alone and I think everything else should fall into place.

Do you mean for `YieldType`? We still have a `STAND_ALONE` field in `MtuType`.

> I would pull the queries for NET and for GEN, leaving the LOAD query alone for now until we figure out what it does.

For me the query with `D=0` (currently labeled net in my pr) is the only query that returns nonzero values. Did I mess that up and is the `D=0` query supposed to be the LOAD query? Because otherwise removing `D=0` would break things.

> I would say all Spyder values are Stand-Alone, they aren't used for any of the totaling values. An MTU will be part of the NET total calculation so calling it NET could make sense.

Now that I think about it more, I don't think keeping around `YieldType` is useful, since Spyders and MTU don't fit in well with the load/gen/net/stand-alone definitions. Besides, for every method the `YieldType` is fixed, so you know if you call `.energy()` on a spyder you will get spyder/stand-alone energy, if you call `.consumed()` you will get load energy, if you call`.energy()` on an MTU you either get NET energy (if MTUType=NET) or stand-alone (if MTUType=Standalone), etc.

**Comment by realumhelp** [comment]
I agree it might be a stretch to keep YieldType since it only really make sense for the DashData queries. Any individual queries (ie:MTU/Spyder) it doesn't really mean anything if you remove the MTU and Spyder types. 

DashData.xml Query 'D' params:
D=0 is for NET
D=1 is for LOAD
D=2 is for Generation

If your system is a LOAD-GEN type I don't think we can support this without getting some system behavior. I have high confidence that if your system is NET-GEN then both the NET and the Generation queries on DashData.xml will behave as expect. That's what I was trying to get across.

**Comment by realumhelp** [comment]
My current guess is that if your system is LOAD-GEN then the DashData.xml NET queries will return all 0's and the LOAD queries will return the cumulative values.

**Comment by realumhelp** [comment]
You could probably boil the system down into two different types: 
NET + GEN and LOAD + GEN

If you only have one MTU you fall into the NET + GEN category but the GEN queries will always return 0.

**Comment by rianadon** [comment]
It sounds like without having a type besides NET for testing, it'll be difficult to correctly implement the other system types. For now I'm thinking we could check when retrieving net if the yield = `EnergyYield(0, 0, 0)`, and if so return the load so that we can always send the net value to Home Assistant and if the system type has + GEN then return net as well as gen?

**Comment by realumhelp** [comment]
> It sounds like without having a type besides NET for testing, it'll be difficult to correctly implement the other system types. For now I'm thinking we could check when retrieving net if the yield = `EnergyYield(0, 0, 0)`, and if so return the load so that we can always send the net value to Home Assistant and if the system type has + GEN then return net as well as gen?

Definitely drawing straws in the dark. So either compare against 0 results or:
```
energy():
  if system_type == SystemType.NET or system_type == SystemType.NET_GEN:
    return netqueryresults
  else:
    return loadqueryresults
```

It should be safe to always return the generation query. If you don't have any generation capability it will just always return 0.

**Comment by realumhelp** [comment]
I just factory reset my system for science to try and see what this LOAD business is! 

Test #1 - Setup LOAD-GEN using my normal MTU as a LOAD, fake MTU as GEN:
 - DashData.xml - NET:
   - Returned what looks like the NET of the system.
 - DashData.xml - LOAD:
   - Returned what looked like the same as the NET query.
 - DashData.xml - GEN:
   - Returned all 0's as expected.

Test #2 - Setup LOAD-GEN using my normal MTU as a GEN, fake MTU as LOAD:
 - DashData.xml - NET:
   - Returned what looks like the NET of the system.
 - DashData.xml - LOAD:
   - Returned all 0's
 - DashData.xml - GEN:
   - Returned what looks like the NET of the system. It wasn't a negative value which was weird maybe??

Test #3 - Setup NET-GEN using my normal MTU as a NET, fake MTU as GEN:
 - DashData.xml - NET:
   - Returned what looks like the NET of the system.
 - DashData.xml - LOAD:
   - Returned what looked like the same as the NET query.
 - DashData.xml - GEN:
   - Returned all 0's as expected.
   
 Test #4 - Setup NET-GEN using my normal MTU as a GEN, fake MTU as NET:
 - DashData.xml - NET:
   - Returned all 0's
 - DashData.xml - LOAD:
   - Returned what looks like the GEN of the system but NEGATIVE!!
 - DashData.xml - GEN:
   - Returned what looks like the NET of the system. It wasn't a negative value which was weird maybe??

 Test #5 - Setup NET using my normal MTU as a NET:
 - DashData.xml - NET:
   - Returned what looks like the NET of the system.
 - DashData.xml - LOAD:
   - Returned all 0's
 - DashData.xml - GEN:
   - Returned all 0's

So mind boggled but those are the results.

**Comment by realumhelp** [comment]
I ran Test 2 again because the results looked weird (MTU1 is fake and MTU2 is real):

```
<Configuration>
  <SystemType>2</SystemType>
  <MTUType1>1</MTUType1>
  <MTUType2>2</MTUType2>
  <MTUType3>0</MTUType3>
  <MTUType4>0</MTUType4>
</Configuration>
```


api/DashData.xml?T=0&D=0&M=0
```
<DashData>
  <Now>2088</Now>
  <TDY>270</TDY>
  <MTD>270</MTD>
  <Avg>42480</Avg>
  <Proj>1316880</Proj>
  <Voltage>0</Voltage>
  <Phase>195</Phase>
</DashData>
```

api/DashData.xml?T=0&D=1&M=0
```
<DashData>
  <Now>0</Now>
  <TDY>0</TDY>
  <MTD>0</MTD>
  <Avg>0</Avg>
  <Proj>0</Proj>
  <Voltage>0</Voltage>
  <Phase>195</Phase>
</DashData>
```

api/DashData.xml?T=0&D=2&M=0
```
<DashData>
  <Now>2086</Now>
  <TDY>270</TDY>
  <MTD>270</MTD>
  <Avg>42480</Avg>
  <Proj>1316880</Proj>
  <Voltage>0</Voltage>
  <Phase>195</Phase>
</DashData>
```

**Comment by rianadon** [comment]
> I just factory reset my system for science to try and see what this LOAD business is!

Wow. Thank you for doing this! I admire the dedication.

As for GEN being positive, I think this might be the solution (from the [installation manual](https://www.theenergydetective.com/downloads/TED%20Pro%20Home-Lite%20Installation%204.0.pdf))
image

**Comment by realumhelp** [comment]
Yea that is probably the issue. I'll see if I get some time to open my panel up and see about switching it. I'm not sure it will make much of a difference anyway since I'm not sure the impact the fake MTU had on things.

I think the original plan to check for NET == EnergyYield(0, 0, 0) is still a good path forward.

**Comment by rianadon** [comment]
I made a few changes:
- In `.consumption()`, the sign of `GEN` gets flipped (since `GEN` should be negative, which matches up with test data).
- I added [a section in the README](https://github.com/rianadon/the-energy-detective-py/tree/api-change#notes) on system types summarizing the conversation so far and what you found.
- I decided against doing the `NET == EnergyYield(0, 0, 0)` check because while creating the table I realized that for NET and NET_GEN, there will always be a NET MTU connected, and if that MTU is returning zero that only means that the home consumption equals the solar production. Based on your testing, it looks like LOAD_GEN correctly calculates NET.
- However, `.consumption()` for NET type grabs the NET values rather than the LOAD values since NET doesn't publish LOAD values.

I also noticed you used LOAD-GEN and NET-LOAD interchangeably. I think LOAD-GEN is the correct term to use, right?

**Comment by rianadon** [comment]
Also thank you again for your testing! I think I am finally understanding how the API works with your help.

**Comment by realumhelp** [comment]
Yep typo sorry. NET-LOAD doesn't exist it is LOAD-GEN. Even after all those tests I don't think I could 100% tell you exactly what NET/LOAD/GEN all mean. The interface isn't very self documenting via reverse engineering. 

Just a note on setting the generation negative you might want to do the absolute value just in case the value itself if actually negative. You don't want to accidentally flip it positive because as you said in the TED instructions I may have to flip some wires around to get a negative reading.

**Comment by rianadon** [comment]
> Just a note on setting the generation negative you might want to do the absolute value just in case the value itself if actually negative. You don't want to accidentally flip it positive because as you said in the TED instructions I may have to flip some wires around to get a negative reading.

The problem is that NET and/or LOAD will also depend on how the magnitude of GEN, so if GEN Is plugged in incorrectly then there will be more problems than just GEN being negative. So I think it's better to keep GEN negative if it's plugged in wrong so that it's easier to notice there's a problem.

File: README.md
Comment (realumhelp) [inline README.md]: 'likeley' spelling.

File: src/tedpy/ted5000.py
Comment (realumhelp) [inline src/tedpy/ted5000.py:73]: I wonder if we could be smart and total the MTU's for atleast generation(). Since the TED5000 interface has a bit more information in it I think we could do it fairly easily by looping over the MTUs. 

```
generation():
  for mtu in mtus:
    if(int(data["Power"]["MTU%d" % mtu.position]["PowerNow"]) < 0):
      total_gen += int(data["Power"]["MTU%d" % mtu.position]["PowerNow"])
```

You could do the same for Now, TDY, and MTD, if you thought it had value added.

```
        power_now = int(data["Power"]["MTU%d" % mtu.position]["PowerNow"])
        power_tdy = int(data["Power"]["MTU%d" % mtu.position]["PowerTDY"])
        power_mtd = int(data["Power"]["MTU%d" % mtu.position]["PowerMTD"])
```

File: src/tedpy/ted6000.py
Comment (realumhelp) [inline src/tedpy/ted6000.py:115]: From previous comments I assume you are expecting production to return a positive value? The instructions state that the production returned by TED should be negative when properly configured. When I was doing my testing above my system was returning positive values so it was not properly configured for production.

Comment (realumhelp) [reply src/tedpy/ted6000.py:115]: I see from the unit tests that production is expected to return positive and the MTU itself is going to return negative.

**Review by realumhelp — APPROVED** [review]
Not that you need my approval but I looked through the changes and they look good. I added a few individual comments/questions/suggestions.

File: src/tedpy/dataclasses.py
Comment (realumhelp) [inline src/tedpy/dataclasses.py:58]: I did some digging on this one because I was wondering how I made such a bad typo. It turns out the actual interface spec calls out "NET+LOAD" as the proper name. I got the LOAD_GEN stuff from a forum post on the TED forums mixing TED5000 threads in TED6000 code. Probably should revert this back to "NET_LOAD"

File: src/tedpy/ted5000.py
Comment (realumhelp) [reply src/tedpy/ted5000.py:73]: A bit better example of what I was trying to get across. Not exactly Python code but should get the idea across. I believe you want production to return a positive value. Otherwise remove the abs().

```python
    def production(self) -> EnergyYield:
        """Return generation information for the whole system."""
        gen = EnergyYield(0, 0, 0)
        for mtu in mtus:
          e = self._mtu_energy(mtu)
          if(min(e.now, e.daily, e.mtd) < 0):
            gen.now += abs(e.now)
            gen.daily += abs(e.daily)
            gen.mtd += abs(e.mtd)
        return gen
```

The idea being that gen.now and gen.daily could be 0 if it was a cloudy day or something silly like that. Most likely the monthly will be negative, unless it is the beginning of the month. In which case ignore the 0 values and catch them next update.

Comment (rianadon) [reply src/tedpy/ted5000.py:73]: I'm a bit confused how this would be helpful. Shouldn't the gen totals read from the API already be totalling the MTU? What would looping over the MTUs give us?

File: README.md
Comment (rianadon) [reply README.md]: Thanks!

File: src/tedpy/ted6000.py
Comment (rianadon) [reply src/tedpy/ted6000.py:115]: Yes; the MTU should return negative and the production() method should return positive.

Now I realize I should probably do a check on the MTU type and make sure MTUs configured as GEN return positive as well.

File: src/tedpy/dataclasses.py
Comment (rianadon) [reply src/tedpy/dataclasses.py:58]: I like the `LOAD_GEN` name better though, because it seems in that mode you can only configure MTUs as GEN or LOAD, not NET. So I don't know why they chose the name NET+LOAD.

File: src/tedpy/ted5000.py
Comment (realumhelp) [reply src/tedpy/ted5000.py:73]: > I'm a bit confused how this would be helpful. Shouldn't the gen totals read from the API already be totalling the MTU? What would looping over the MTUs give us?

This is for the TED5000 which does not have the DashData queries like the TED6000 does. In an effort to make a top level production() function for the TED5000 summing the values is an idea. Although now I see there is an MTU Type so that would be better than using the mtu_energy method.

`if mtu.type == MtuType.GENERATION:`

File: src/tedpy/dataclasses.py
Comment (realumhelp) [reply src/tedpy/dataclasses.py:58]: Agreed the name doesn't make much sense. I'm fine with keeping it as is.

File: src/tedpy/ted6000.py
Comment (realumhelp) [inline src/tedpy/ted6000.py:103]: Consider making a similar change on the TED5000 side so any generation there returns positive as well.

**Comment by rianadon** [comment]
I edited the TED5000 `_parse_mtu_type`, as according to the documentation type zero is a LOAD: 

image

[This forum post](http://forum.theenergydetective.com/index.php?topic=202.0) says that Adj. Load is bi-directional, so is equivalent to NET.

I also implemented the TED5000 `.consumption()` method as summing up LOAD mtus and the `.production()` method as summing GEN mtus. I could also try to include the stand-alone and the net mtus depending on whether they are positive and negative, but this doesn't make sense for cases where the today's energy is positive but monthly energy is negative for example.

**Comment by realumhelp** [comment]
Good catch!

**Comment by rianadon** [comment]
I'm going to merge this one in. Thank you for all the reviewing!

**Pull request merged**
